src/librustc_interface/queries.rs

```diff
@@ -278,35 +276,54 @@ impl<'comp> Queries<'comp> {
         })
     }
 
-    pub fn link(&self) -> Result<&Query<()>> {
-        self.link.compute(|| {
-            let sess = self.session();
+    pub fn linker(self) -> Result<Linker> {
+        let dep_graph = self.dep_graph()?;
+        let prepare_outputs = self.prepare_outputs()?;
+        let ongoing_codegen = self.ongoing_codegen()?;
 
-            let ongoing_codegen = self.ongoing_codegen()?.take();
+        let sess = self.session().clone();
+        let codegen_backend = self.codegen_backend().clone();
 
-            self.codegen_backend().join_codegen_and_link(
-                ongoing_codegen,
-                sess,
-                &*self.dep_graph()?.peek(),
-                &*self.prepare_outputs()?.peek(),
-            ).map_err(|_| ErrorReported)?;
-
-            Ok(())
+        Ok(Linker {
+            sess,
+            dep_graph: dep_graph.take(),
```

Comment:
The `DepGraph` has a `Rc` internally so you could just clone it instead with `.peek().clone()`.